### PR TITLE
Instantiate test data object via an annotation #90

### PR DESCRIPTION
**Pull request by oomelianchuk**
Please review the changes and if you agree with the implementation I will add tests for this.

The changes are meant to fulfill [this](https://github.com/Xceptance/neodymium-library/issues/90) ticket. I also added an option to select the JSON field from which the object data will be retreived

### CODE REVIEW

File: src/test/java/com/xceptance/neodymium/tests/TestDataStatementTest.java [hunk 1]
```diff
@@ -267,10 +268,9 @@ public void testMethodDefaultTwoDataSet() throws Throwable
     {
         // Two data sets and @Testdata on method
         String[] expected = new String[]
-        {
-          "test1 :: Data set 1 / 2",
-          "test1 :: Data set 2 / 2"
-        };
+            {
+              "test1 :: Data set 1 / 2", "test1 :: Data set 2 / 2"
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/tests/TestDataStatementTest.java:272]: Was this change intended needed? Can we go back to the old format for better readability.

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import org.junit.runner.RunWith;
+
+import com.xceptance.neodymium.NeodymiumRunner;
+import com.xceptance.neodymium.module.statement.testdata.Data;
+import com.xceptance.neodymium.util.DataUtils;
+
+@RunWith(NeodymiumRunner.class)
+public class InstantiateDtoViaJsonPathInAnnotation
+{
+    @Data("$.user")
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java:14]: Could you add another test retrieving an object from a more complex structure?

e.g `"$.user.payment.creditcards[2]"`

Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java:14]: To avoid introducing another data structure I used ` "$.user.contacts.friends[2]"`. Done

File: src/test/java/com/xceptance/neodymium/tests/TestDataStatementTest.java (hunk 1 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/tests/TestDataStatementTest.java]: Formatting changes reverted

File: src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java [hunk 3]
```diff
@@ -64,9 +70,50 @@ public TestdataStatement()
...
+    private void initializeDataObjects() throws IllegalArgumentException, IllegalAccessException
+    {
+        for (Field field :getFieldsFromSuperclasses())
+        {
+            Data dataAnnotation = field.getAnnotation(Data.class);
+            if (dataAnnotation != null)
+            {
+                boolean isFieldAccessable = field.canAccess(testClassInstance);
+                field.setAccessible(true);
+                try
```
Comment (occupant23) [inline src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java:86]: I tested the feature and would like to provide an easy access to the first level of objects. Please adjust the code and the test cases accordingly to the following snippet.

```
                try
                {
                    if (!StringUtils.isBlank(dataAnnotation.value()))
                    {
                        field.set(testClassInstance, DataUtils.get(dataAnnotation.value(), field.getType()));
                    }
                    else
                    {
                        if (DataUtils.exists(field.getName()))
                        {
                            field.set(testClassInstance, DataUtils.get("$." + field.getName(), field.getType()));
                        }
                        else
                        {
                            field.set(testClassInstance, DataUtils.get(field.getType()));
                        }
                    }
                } catch (Exception e) {
                    // TODO: throw a RuntimeException that contains a hint how this  error can be taken care of
                }
                finally
                {
                    field.setAccessible(isFieldAccessable);
                }
```

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/inheritance/ChildInheritingDtoFromAnnotation.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.xceptance.neodymium.testclasses.data.annotation.inheritance;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import com.xceptance.neodymium.util.DataUtils;
+
+public class ChildInheritingDtoFromAnnotation extends ParentClassWithDtoFromAnnotation
+{
+    @Test
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/data/annotation/inheritance/ChildInheritingDtoFromAnnotation.java:10]: You can use the new feature to optimize your code :-)
Just instantiate the testId as field and use it within the functions.

```
    @Data("$.testId")
    private String testId;
```
after the change above should the following work:
```
    @Data
    private String testId;
```

File: src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java (hunk 3 above)
Comment (occupant23) [reply src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java:86]: We will also need new tests regarding the first level object support.

**Review by occupant23 — CHANGES_REQUESTED** [review]

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/inheritance/ChildInheritingDtoFromAnnotation.java (hunk 4 above)
Comment (occupant23) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/inheritance/ChildInheritingDtoFromAnnotation.java:10]: Btw. Please add this change to all test classes of this feature if it optimizes the code.

Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/inheritance/ChildInheritingDtoFromAnnotation.java:13]: Done

File: src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java (hunk 3 above)
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java:86]: Done. I haven't created a new test for it as I think, that the `InstantiateFieldViaAnnotation` is suitable for this. So I adjusted this test to use field name instead of JSON path. The JSON path is already tested in the `InstantiateDtoViaJsonPathInAnnotation` anyway.

File: src/test/java/com/xceptance/neodymium/tests/TestDataStatementTest.java [hunk 5]
```diff
@@ -452,4 +456,32 @@ public void testRandomDataSetsException()
...
+    public void canInstantiateFieldViaAnnotation()
+    {
+        Result result = JUnitCore.runClasses(InstantiateFieldViaAnnotation.class);
+        checkPass(result, 2, 0);
+    }
+
+    @Test
+    public void canInstantiateDtoViaJsonPathInAnnotation()
+    {
+        Result result = JUnitCore.runClasses(InstantiateDtoViaJsonPathInAnnotation.class);
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/tests/TestDataStatementTest.java:470]: Done

File: src/main/java/com/xceptance/neodymium/module/statement/testdata/Data.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+@Retention(RUNTIME)
+@Target(
+{
+  ElementType.FIELD
+})
+public @interface Data
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/testdata/Data.java]: Done

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+@RunWith(NeodymiumRunner.class)
+public class InstantiateFieldViaAnnotation
+{
+    @DataItem
+    private String name;
+
+    @DataItem
+    private String testId;
+
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java:18]: add filed to validate the default values

```
    
    @DataItem
    private String stringDefaultValue;

    @DataItem
    private int intDefaultValue;

    @DataItem
    private long longDefaultValue;

    @DataItem
    private float floatDefaultValue;

    @DataItem
    private double doubleDefaultValue;

    @DataItem
    private boolean booleanDefaultValue;
```

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    @Test
+    public void test1()
+    {
+        Assert.assertEquals("John" + testId, name);
+    }
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java:23]: call `validateDataItemDefaultValues();`

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        Assert.assertEquals("john" + DataUtils.asString("testId") + "@varmail.de", user.getEmail());
+        Assert.assertEquals("neodymium" + DataUtils.asString("testId"), user.getPassword());
+    }
+
+    @Test
+    public void test2()
+    {
+        Assert.assertEquals("friend.of.john" + DataUtils.asString("testId") + "@varmail.de", friend.getEmail());
+        Assert.assertEquals("neodymium_friend_of_john" + DataUtils.asString("testId"), friend.getPassword());
+    }
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java:32]: call `validateDataItemDefaultValues();`

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+    @DataItem("$.user.contacts.friends[2]")
+    private User friend;
+
+    @Test
+    public void test1()
+    {
+        Assert.assertEquals("john" + DataUtils.asString("testId") + "@varmail.de", user.getEmail());
+        Assert.assertEquals("neodymium" + DataUtils.asString("testId"), user.getPassword());
+    }
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java:25]: call `validateDataItemDefaultValues();`

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import com.xceptance.neodymium.NeodymiumRunner;
+import com.xceptance.neodymium.module.statement.testdata.DataItem;
+import com.xceptance.neodymium.util.DataUtils;
+
+@RunWith(NeodymiumRunner.class)
+public class InstantiateDtoViaJsonPathInAnnotation
```
Comment (occupant23) [inline src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java:12]: extend `InstantiateFieldViaAnnotation`

**Review by occupant23 — CHANGES_REQUESTED** [review]
improve tests and error message

File: src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java [hunk 12]
```diff
@@ -64,9 +70,58 @@ public TestdataStatement()
...
+                        field.set(testClassInstance, DataUtils.get("$." + field.getName(), field.getType()));
+                    }
+                    else if (DataUtils.getDataAsJsonObject().isJsonPrimitive() ==( field.getType().isPrimitive()|| field.getType().equals(String.class)) )
+                    {
+                        field.set(testClassInstance, DataUtils.get(field.getType()));
+                    }
+                }
+                catch (Exception e)
+                {
+                    throw new RuntimeException("Something went wrong while test data value injection", e);
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/main/java/com/xceptance/neodymium/module/statement/testdata/TestdataStatement.java]: Done

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+    public void test1()
+    {
+        Assert.assertEquals("john" + DataUtils.asString("testId") + "@varmail.de", user.getEmail());
+        Assert.assertEquals("neodymium" + DataUtils.asString("testId"), user.getPassword());
+    }
+
+    @Test
+    public void test2()
+    {
+        Assert.assertEquals("friend.of.john" + DataUtils.asString("testId") + "@varmail.de", friend.getEmail());
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java]: Done

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    @DataItem("$.user")
+    private User user;
+
+    @DataItem("$.user.contacts.friends[2]")
+    private User friend;
+
+    @Test
+    public void test1()
+    {
+        Assert.assertEquals("john" + DataUtils.asString("testId") + "@varmail.de", user.getEmail());
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java]: Done

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java (hunk 11 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateDtoViaJsonPathInAnnotation.java]: Done

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java (hunk 8 above)
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java:67]: Done

Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java:67]: Done

File: src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+}
```
[the remark this answers is not in the document]
Comment (oomelianchuk) [reply src/test/java/com/xceptance/neodymium/testclasses/data/annotation/InstantiateFieldViaAnnotation.java:68]: Done

**Comment by wurzelkuchen** [comment]
@oomelianchuk Can you add some JavaDoc where it's usefull?

**Comment by oomelianchuk** [comment]
Sure! Javadoc for the `@DataItem` annotation is added.

**Review by wurzelkuchen — APPROVED** [review]
Looks all good now
